share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java

```diff
@@ -0,0 +1,536 @@
+                                        .setLeaderEpoch(partitionData.leaderEpoch())
+                                        .setStateEpoch(partitionData.stateEpoch())
+                                        .setStateBatches(partitionData.stateBatches())))))))
+                        .exceptionally(exception -> handleOperationException(
+                            "write-share-group-state",
+                            request,
+                            exception,
+                            (error, message) -> WriteShareGroupStateResponse.toErrorResponseData(
+                                topicData.topicId(),
+                                partitionData.partition(),
+                                error,
+                                "Unable to write share group state: " + exception.getMessage()
+                            ),
+                            log
+                        ));
+                    partitionFut.put(partitionData.partition(), future);
+                });
+        });
+
+        // Combine all futures into a single CompletableFuture<Void>
+        CompletableFuture<Void> combinedFuture = CompletableFuture.allOf(futureMap.values().stream()
+            .flatMap(partMap -> partMap.values().stream()).toArray(CompletableFuture[]::new));
+
+        // topicId -> {partitionId -> responseFuture}
+        return combinedFuture.thenApply(v -> {
+            List<WriteShareGroupStateResponseData.WriteStateResult> writeStateResults = new ArrayList<>(futureMap.size());
+            futureMap.forEach(
+                (topicId, topicEntry) -> {
+                    List<WriteShareGroupStateResponseData.PartitionResult> partitionResults = new ArrayList<>(topicEntry.size());
+                    topicEntry.forEach(
+                        // map of partition id -> responses from api
+                        (partitionId, responseFut) -> {
+                            // This is the future returned by runtime.scheduleWriteOperation which returns when the
+                            // operation has completed including error information. When this line executes, the future
+                            // should be complete as we used CompletableFuture::allOf to get a combined future from
+                            // all futures in the map.
+                            WriteShareGroupStateResponseData partitionData = responseFut.getNow(null);
+                            partitionResults.addAll(partitionData.results().get(0).partitions());
+                        }
+                    );
+                    writeStateResults.add(WriteShareGroupStateResponse.toResponseWriteStateResult(topicId, partitionResults));
+                }
+            );
+
+            // time taken for write
+            shareCoordinatorMetrics.record(ShareCoordinatorMetrics.SHARE_COORDINATOR_WRITE_LATENCY_SENSOR_NAME,
+                time.hiResClockMs() - startTime);
```

Comment:
nit: you have defined `startTime` above. I think we should use that here.

Reply:
This is being used to calculate the delta 
we want to record how much time it took for the writeState call.